File: job-challenges-code/leetcode_solutions/DeepCopyOfLinkedList/DeepCopyOfLinkedList.java
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
 
        HashMap<Node, Node> oldToCopy = new HashMap<>();
        Node current = head;

        while (current != null) {
            Node copy = new Node(current.val);
            oldToCopy.put(current, copy);
            current = current.next;
        }

        current = head;

        while (current != null) {
            Node copy = oldToCopy.get(current);
            copy.next = oldToCopy.get(current.next);
            copy.random = oldToCopy.get(current.random);
            current = current.next;
        }
        
        System.out.println(oldToCopy.get(head));

        return oldToCopy.get(head);
    }
}

File: job-challenges-code/leetcode_solutions/DeepCopyOfLinkedList/DeepCopyOfLinkedList.py
"""
# Definition for node
class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Node") -> "Node":
        oldToCopy = {}
        
        current = head
        while current:
            copy = Node(current.val)
            oldToCopy[cur] = copy
            current = current.next
            
        
        current = head
        
        while current:
            copy = oldToCopy[current]
            copy.next = oldToCopy[current.next]
            copy.random = oldToCopy[current.random]
            current = current.next
            
        return oldToCopy[head]